File: BindingFilter.pl
# Based on myParseReportsTopGenesTopAlleleStringent_4.pl by TWYLIE (Feb 2012)
#Takes in a FOF with parsed NetMHC files for different allele-length combinations and outputs best candidates per gene based on binding affinity.
use strict;
use warnings;
use File::Basename qw(basename);
use Getopt::Std;


use vars qw($opt_i $opt_f $opt_o $opt_c $opt_b);
getopts('i:f:o:c:b:');


if(! $opt_f || ! $opt_i || ! $opt_o || ! $opt_c || ! $opt_b){

   print "Usage: $0\n";
   print "-i <input list of variants>\n";
   print "-f <FOF containing list of parsed epitope files for different allele-length combinations (same sample)>\n";
   print "-o <Output .xls file containing list of filtered epitopes based on binding affinity for each allele-length combination per gene>\n";
   print "-c <Minimum fold change between mutant binding score and wild-type score. The default is 0, which filters no results, but 1 is often a sensible default 
   (requiring that binding is better to the MT than WT)>\n";
   print "-b <report only epitopes where the mutant allele has ic50 binding scores below this value ; default 500>\n";

	die "ERROR : Please make sure all parameters are defined\n";
}

my $variant_file = $opt_i;
my $fof_file = $opt_f;
my $output_file = $opt_o;
my $binding_threshold = $opt_b || 500;
my $minimum_fold_change = $opt_c || 0;


#$binding_threshold = 500;
#$minimum_fold_change = 0;

if(! -e $variant_file){
   die "Variant File $variant_file doesn't exist -- fatal.\n";
}


    my %variants;
    open(IFH,$variant_file) || die "Can't open $variant_file.\n";
    while (my $line = <IFH>) {
            chomp $line;
            #header?
            if($line=~/^chromosome_name/){
                $variants{"header"} = $line;
                next;
            }
            my @F = split("\t",$line); 
            #1	chromosome_name
			#2	start
			#3	stop
			#4	reference
			#5	variant
			#6	gene_name
			#7	transcript_name
			#8	amino_acid_change
			#9	ensembl_gene_id
			#10	wildtype_amino_acid_sequence
                       
            my $gene = $F[5];
            my $amino_acid_change = $F[7];
            $variants{join("\t",($gene,$amino_acid_change))} = $line;
        }
       if(defined($variants{"header"})){
		open(OUT,">$output_file") || die "Can't open $output_file for writing.\n";	
            print OUT join("\t",($variants{"header"},
                                     "GeneName","MutationEffect","HLAallele","PeptideLength",
                                     'SubPeptidePosition','MT Score',
                                     'WT Score','MT Epitope Seq',
                                     'WT Epitope Seq','Fold Change')) . "\n";
        }

    # read in the netMHC predictions and filter
        open(FOF,$fof_file) || die "Can't open $fof_file.\n";


    my %prediction;
    my $threshold = $binding_threshold;
    while (my $file = <FOF>) {
        chomp $file;
        my $basename = basename( $file);
        # TEST.HLA-A01:01.9.netmhcM.parsed
        my @f      = split( /\./, $basename );
        my $sample = $f[0];
       $sample =~ s/_netmhc//g;
        my $allele = $f[1];
        my $length = $f[2];

        my $mode = 'filtered';
        my $i = 0;
        
        open(FILE,$file)
  		or die "Could not open file '$file' $!";
 		FILELINE:
   		while (<FILE>) {
        chomp;
        my $line = $_;
        $i++;
        next if ($i == 1);  # skips headers....
        my (
            $gene_name,
            $point_mutation,
            $sub_peptide_position,
            $mt_score,
            $wt_score,
            $mt_epitope_seq,
            $wt_epitope_seq,
            $fold_change,
           ) = split( "\t", $line );


        $gene_name = {
                      gene_name            => $gene_name,
                      allele               => $allele,
                      point_mutation       => $point_mutation,
                      sub_peptide_position => $sub_peptide_position,
                      mt_score             => $mt_score,
                      wt_score             => $wt_score,
                      mt_epitope_seq       => $mt_epitope_seq,
                      wt_epitope_seq       => $wt_epitope_seq,
                      fold_change          => $fold_change,
                     };
        push( @{ $prediction{$mode}->{$sample}->{$length}->{genes} }, \$gene_name );
    }
    

}
close (FOF);


##

my %best;

MODE:
foreach my $mode (sort keys %prediction) {
    SAMPLE:
    foreach my $sample (sort keys %{ $prediction{$mode} }) {
       LENGTH:
        foreach my $length (sort keys %{ $prediction{$mode}->{$sample} }) {
            GENES:
            foreach my $gene (sort @{ $prediction{$mode}->{$sample}->{$length}->{genes} }) {
                # BEST
                unless( !$best{$sample}->{$$gene->{gene_name}}->{SCORE} ) {
                    if ($$gene->{mt_score} < $best{$sample}->{$$gene->{gene_name}}->{SCORE}) {
                        $best{$sample}->{$$gene->{gene_name}}->{SCORE} = $$gene->{mt_score};
                        $best{$sample}->{$$gene->{gene_name}}->{GENES} = [];
                        $$gene->{sample} = $sample;
                        $$gene->{length} = $length;
                        $$gene->{mode}   = $mode;
                        push( @{ $best{$sample}->{$$gene->{gene_name}}->{GENES} }, $gene );
                    }
                    elsif ($$gene->{mt_score} == $best{$sample}->{$$gene->{gene_name}}->{SCORE}) {
                        $best{$sample}->{$$gene->{gene_name}}->{SCORE} = $$gene->{mt_score};
                        $$gene->{sample} = $sample;
                        $$gene->{length} = $length;
                        $$gene->{mode}   = $mode;
                        push( @{ $best{$sample}->{$$gene->{gene_name}}->{GENES} }, $gene );
                    }
                }
                else {
                    $best{$sample}->{$$gene->{gene_name}}->{SCORE} = $$gene->{mt_score};
                    $$gene->{sample} = $sample;
                    $$gene->{length} = $length;
                    $$gene->{mode}   = $mode;
                    push( @{ $best{$sample}->{$$gene->{gene_name}}->{GENES} }, $gene );
                }

            }
        }
    }
}

# REPORTING
    foreach my $sample (sort keys %best) {
        foreach my $gene (sort keys %{ $best{$sample} }) {
            foreach my $entry (@{ $best{$sample}->{$gene}->{GENES} }) {
                if (($$entry->{mt_score} < $threshold) && ($$entry->{fold_change} > $minimum_fold_change)){
                    #write out epiptopes
    #                print OUT $entry;
                    #also write out variants with epitopes appended
                    if(defined($variant_file)){
                        my $key = join("\t",$gene,$$entry->{point_mutation});
                        if(defined($variants{$key})){                            
                            print OUT join("\t", ($variants{$key},
                                                      $gene,$$entry->{point_mutation},$$entry->{allele},
                                                      $$entry->{length},$$entry->{sub_peptide_position},$$entry->{mt_score},
                                                      $$entry->{wt_score},$$entry->{mt_epitope_seq},
                                                      $$entry->{wt_epitope_seq},$$entry->{fold_change})) . "\n";

                        } else {
                            
                            print "Couldn't find variant for " . $gene . " " . $$entry->{point_mutation} . "in variant_file";
                        }
                    }
                }
            }
        }
    }

close (OUT);
close (IFH);

File: pvac_seq/binding_filter.py
import argparse
import sys
import re
import os


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('-i', '--input', type=argparse.FileType('r'),
                        help="Input list of variants",
                        required=True)
    parser.add_argument('-f', '--fof', type=argparse.FileType('r'),
                        help="OF containing list of parsed epitope files " +
                        "for different allele-length combinations (same sample)",
                        required=True)
    parser.add_argument('-o', '--output', type=argparse.FileType('w'),
                        help="Output .xls file containing list of filtered " +
                        "epitopes based on binding affinity for each " +
                        "allele-length combination per gene",
                        required=True)
    parser.add_argument('-c', '--fold-change', type=int,
                        help="Minimum fold change between mutant binding " +
                        "score and wild-type score. The default is 0, which " +
                        "filters no results, but 1 is often a sensible " +
                        "default (requiring that binding is better to the MT " +
                        "than WT)",
                        default=0,
                        required=False,
                        dest="minimum_fold_change")
    parser.add_argument('-b', '--binding-threshold', type=int,
                        help="Report only epitopes where the mutant allele " +
                        "has ic50 binding scores below this value; default 500",
                        default=500,
                        required=False,
                        dest="binding_threshold")

    args = parser.parse_args()

    #precompile regex patterns used later
    chromosome_name = re.compile(r'^chromosome_name')
    netmhc_subber = re.compile(r"_netmhc")

    #open the variants file and parse into variants dictionary
    intake = args.input.readline().rstrip()
    variants = {}

    while intake != '':
        if chromosome_name.match(intake):
            variants['header'] = intake
            intake = args.input.readline().rstrip()
            continue

        data = intake.split('\t')

        #1	chromosome_name
        #2	start
        #3	stop
        #4	reference
        #5	variant
        #6	gene_name
        #7	transcript_name
        #8	amino_acid_change
        #9	ensembl_gene_id
        #10	wildtype_amino_acid_sequence

        gene = data[5]
        amino_acid_change = data[7]
        variants[ gene + '\t' + amino_acid_change] = intake

        intake = args.input.readline().rstrip()

    #dump header data to output file
    args.input.close()
    if 'header' not in variants:
        # usage("Header not defined in variant input file")
        parser.print_help()
        print("\nError: Header not defined in variant input file")
        exit(1)
    args.output.write('\t'.join([
        variants['header'], "GeneName", "HLAallele",
        "PeptideLength", "SubPeptidePosition",
        "MTScore", "WTScore", "MTEpitopeSeq",
        "WTEpitopeSeq", "FoldChange"
        ]))
    args.output.write("\n")

    #Read netmhc files from the fof, and parse into predictions
    prediction = {}

    intake = args.fof.readline().rstrip()


    while intake != "":

        #parse the filepath
        basename = os.path.basename(intake)
        data = basename.split(".")
        sample = netmhc_subber.sub("", data[0])
        allele = data[1]
        length = data[2]
        mode = 'filtered'

        #open each file listed in the fof, and read gene data
        netmhc_reader = open(intake, mode = 'r')
        netmhc_reader.readline() #skip first line
        netmhc_intake = netmhc_reader.readline().rstrip()
        while netmhc_intake != "":

            line_data = netmhc_intake.split("\t")[:8]

            gene_name = {
                'gene_name' : line_data[0],
                'allele' : allele,
                'point_mutation' : line_data[1],
                'sub_peptide_position' : line_data[2],
                'mt_score' : line_data[3],
                'wt_score' : line_data[4],
                'mt_epitope_seq' : line_data[5],
                'wt_epitope_seq' : line_data[6],
                'fold_change' : line_data[7]
                }

            #create the nested dictionary structure if necessary
            if mode not in prediction:
                prediction[mode] = {}
            if sample not in prediction[mode]:
                prediction[mode][sample] = {}
            if length not in prediction[mode][sample]:
                prediction[mode][sample][length] = {}
            if 'genes' not in prediction[mode][sample][length]:
                prediction[mode][sample][length]['genes']=[]

            prediction[mode][sample][length]['genes'].append(gene_name)
            netmhc_intake = netmhc_reader.readline().rstrip()
        netmhc_reader.close()

        intake = args.fof.readline().rstrip()
    args.fof.close()

    best = {}
    #now select the best predictions
    for mode in sorted(prediction):
        for sample in sorted(prediction[mode]):
            for length in sorted(prediction[mode][sample]):
                for gene in prediction[mode][sample][length]['genes']:
                    if (sample in best and
                            gene['gene_name'] in best[sample] and
                            'SCORE' in best[sample][gene['gene_name']]):
                        if (float(gene['mt_score']) <
                                best[sample][gene['gene_name']]['SCORE']):
                            best[sample][gene['gene_name']]['SCORE'] = float(gene['mt_score'])
                            gene['sample'] = sample
                            gene['length'] = length
                            gene['mode'] = mode
                            best[sample][gene['gene_name']]['GENES'] = [gene]
                        elif (float(gene['mt_score']) ==
                                best[sample][gene['gene_name']]['SCORE']):
                            gene['sample'] = sample
                            gene['length'] = length
                            gene['mode'] = mode
                            best[sample][gene['gene_name']]['GENES'].append(gene)
                    else:
                        if sample not in best:
                            best[sample] = {}
                        if gene['gene_name'] not in best[sample]:
                            best[sample][gene['gene_name']] = {}
                        best[sample][gene['gene_name']]['SCORE'] = float(gene['mt_score'])
                        gene['sample'] = sample
                        gene['length'] = length
                        gene['mode'] = mode
                        best[sample][gene['gene_name']]['GENES'] = [gene]

    #Finish off by dumping the selections to the output file
    for sample in sorted(best):
        for gene in sorted(best[sample]):
            for entry in best[sample][gene]['GENES']:
                if (float(entry['mt_score']) < args.binding_threshold and
                        float(entry['fold_change']) > args.minimum_fold_change):
                    key = gene + "\t" + entry['point_mutation']
                    if key in variants:
                        args.output.write("\t".join([
                            variants[key],
                            gene,
                            entry['allele'],
                            entry['length'],
                            entry['sub_peptide_position'],
                            entry['mt_score'],
                            entry['wt_score'],
                            entry['mt_epitope_seq'],
                            entry['wt_epitope_seq'],
                            entry['fold_change']
                            ]))
                        args.output.write("\n")
                    else:
                        print("Couldn't find variant for", gene,
                              entry['point_mutation'], "in variant file")
    args.output.close()


if __name__ == "__main__":
    main()
